### PR TITLE
seventh-part

### PR DESCRIPTION
**Pull request by nikfor**

### CODE REVIEW

File: app/controllers/answers_controller.rb [hunk 1]
```diff
@@ -1,23 +1,49 @@
...
+
+  def show
+  end
 
   def create
-    @answer = @question.answers.new(answer_params)
-    @answer.user = current_user
-    if @answer.save
-      flash.notice = "Ваш ответ успешно создан."
+    if current_user
```
Comment (vkurennov) [inline app/controllers/answers_controller.rb]: А куда делся `before_action :authrnticate_user!` в начале контроллера? Он ведь не просто проверяет наличие current_user, а делает редирект на форму логина или возвращает 401 ошибку для ajax-запросов (что правильно).

Добавь этот фильтр и не придется делать проверку на current_user в экшенах. И вот еще что: от того, что эта проврка есть у тебя не выполнится только тело экшена, а вид-то все равно будет рендерится, а зачем его рендерить, если ничего не произошло?

File: app/controllers/answers_controller.rb [hunk 2]
```diff
@@ -1,23 +1,49 @@
...
+      @answer.user = current_user
+      if @answer.save
+        flash.notice = "Ваш ответ успешно создан."
+      end
+    end
+  end
+
+  def update
+    @question = @answer.question
+    if current_user && current_user.owner_of?(@answer)
```
Comment (vkurennov) [inline app/controllers/answers_controller.rb]: То же самое - добавь фильтр в начало контроллера и убери проверку на current_user

File: app/controllers/answers_controller.rb [hunk 3]
```diff
@@ -1,23 +1,49 @@
...
+        flash.alert = "Где то ошибка, проверьте ответ"
+      end
+    else
+      flash.alert = "Вы не можете редактировать чужой ответ"
     end
   end
 
   def destroy
-    if current_user.owner_of?(@answer)
+    if current_user &&  current_user.owner_of?(@answer)
```
Comment (vkurennov) [inline app/controllers/answers_controller.rb]: И опять дублирование ненужной проверки на current_user

File: app/controllers/answers_controller.rb [hunk 4]
```diff
@@ -1,23 +1,49 @@
...
       @answer.destroy
       flash.notice = "Ответ успешно удален"
     else
       flash.alert = "Вы не можете удалять чужие ответы"
     end
-    redirect_to @answer.question
+  end
+
+  def best
+    if current_user && current_user.owner_of?(@answer.question)
```
Comment (vkurennov) [inline app/controllers/answers_controller.rb]: И здесь дублирование

File: app/models/answer.rb [hunk 5]
```diff
@@ -3,4 +3,14 @@ class Answer < ActiveRecord::Base
   belongs_to :user
 
   validates :body, :user_id, :question_id, presence: true
+
+  def self.default_scope
+    order(best: :desc, created_at: :desc)
+  end
+
+  def best!
+    question.answers.update_all(best: false)
```
Comment (vkurennov) [inline app/models/answer.rb]: Надо обернуть это в [транзакцию](http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html), чтобы изменнения были атомарными. И учти, что в транзакции нужно использовать bang-версии методов (c воскл. знаком), т.к. транзакция откатывается только в случае исключения, но у update_all нет bang-версии.

File: app/views/answers/best.js.erb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+
+$(".best_mark").detach();
+$(".answer_"+<%=@answer.id%>).detach();
+$(".answers_read").prepend("<%= j render 'answer', answer: @answer %>");
```
Comment (vkurennov) [inline app/views/answers/best.js.erb]: Есть более кортоткая запись для рендера: `render @answer`, это то же самое.

File: app/views/answers/create.js.erb [hunk 7]
```diff
@@ -1,3 +1,9 @@
- $(".answers_read").append("<%= j render 'answer' %>");
- $("#answer_body").val("");
- $(".alert").html("<%= flash[:notice] %>");
+<% if @answer.errors.present? %>
+  <% @answer.errors.full_messages.each do |message| %>
+    $("#answer_body").before("<%= message %>")
+  <% end %>
+<% else %>
+  $(".answers_read").append("<%= j render 'answer', answer: @answer %>");
```
Comment (vkurennov) [inline app/views/answers/create.js.erb]: `render @answer`

File: app/views/answers/destroy.js.erb [hunk 8]
```diff
@@ -0,0 +1,2 @@
+$(".answer_"+<%=@answer.id%>).detach();
```
Comment (vkurennov) [inline app/views/answers/destroy.js.erb]: Тут лучше использовать remove, а не detach, т.к. при detach элемент сохраняется в памати. Этот метод подходит, когда элемент повторно нужно вставить в  DOM, но здесь удаление и ничего возвращать не нужно будет.

File: app/views/answers/update.js.erb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<% if @answer.errors.present? %>
+  <% @answer.errors.full_messages.each do |message| %>
+    $(".question_form_errors").html("<%= message %>")
+  <% end %>
+<% else %>
+  $(".answer_"+<%=@answer.id%>).html("<%= j render @answer %>");
```
Comment (vkurennov) [inline app/views/answers/update.js.erb:6]: Вот, здесь же нормальный рендер, чего в других местах так же не написал?

File: app/views/questions/show.html.slim [hunk 10]
```diff
@@ -1,16 +1,21 @@
...
+        = f.button :submit, "Сохранить", class: "right_btn pull-right"
 
   .answers-block
     .answers_read
-      - @question.answers.each do |a|
-        - if a.persisted?
-          .answer.panel.panel-default.clearfix
-            .panel-body = a.body
-            = link_to "Удалить ответ", answer_path(a), class: "btn btn-default", method: :delete if current_user && current_user.owner_of?(a)
+      = render partial: @question.answers
```
Comment (vkurennov) [inline app/views/questions/show.html.slim]: Просто `render @question.answers`, без partial

File: config/routes.rb [hunk 11]
```diff
@@ -4,6 +4,7 @@
     resources :answers, shallow: true
   end
 
+  post 'answers/best/:id', to: 'answers#best', as: :best_answer
```
Comment (vkurennov) [inline config/routes.rb]: Кастомные экшены для ресурсов [задаются по-другому](http://guides.rubyonrails.org/routing.html#adding-more-restful-actions)

File: spec/features/feature_helper.rb [hunk 12]
```diff
@@ -4,6 +4,8 @@
   config.include FeatureMacros, type: :feature
 
   Capybara.javascript_driver = :webkit
+  Capybara.default_max_wait_time = 5
+  Capybara.server_port = 3000
```
Comment (vkurennov) [inline spec/features/feature_helper.rb]: Зачем вешать сервер капибары на тот же порт, где рельсы запускаются? Это не позволит одновременно держать запущенным рельсовый сервер и гонять тесты. Поставь какой-нибудь другой порт, 3030, например.

File: app/views/answers/_answer.html.slim [hunk 13]
```diff
@@ -1,3 +1,15 @@
-.answer.panel.panel-default.clearfix
-  .panel-body = @answer.body
-  = link_to "Удалить ответ", answer_path(@answer), class: "btn btn-default", method: :delete
+- if answer.persisted?
+  div class = "answer_#{answer.id} answer panel panel-default clearfix"
+    - if answer.best
```
Comment (vkurennov) [inline app/views/answers/_answer.html.slim]: Т.к. атрибут best имеет тип boolean, то рельсы автоматом для него создают метод-предикат `best?`и лучше использовать именно его

File: spec/models/answer_spec.rb [hunk 14]
```diff
@@ -5,4 +5,14 @@
...
 
   it { should validate_presence_of :body }
+
+  let!(:question) { create(:question) }
+  let!(:answer) { create(:answer, question: question) }
+  let!(:answer2) { create(:answer, question: question) }
+
+  it "set best field of answer to true" do
+    answer.best!
+    expect(answer.best).to eq true
```
Comment (vkurennov) [inline spec/models/answer_spec.rb]: 1. В юнит-тестах действует правило: один тест-один expect. Поэтому для кажого метода делают отдельный describe и внутри описывают поведение метода в разных ситуациях через несколько тестов
2. Т.к. у нас есть метод-предикат `best?`, то мы можем использовать динамические матчеры рспека:

`expect(answer).to be_best`

**Comment by vkurennov** [comment]
Исправь замечания

File: app/controllers/answers_controller.rb [hunk 15]
```diff
@@ -1,6 +1,10 @@
 class AnswersController < ApplicationController
-  before_action :find_question, except: [:destroy]
-  before_action :find_answer, only: [:destroy]
+  before_action :find_question, except: [:destroy, :update, :best]
+  before_action :find_answer, only: [:destroy, :update, :best]
+  before_action :authenticate_user!
```
Comment (vkurennov) [inline app/controllers/answers_controller.rb]: Этот фильтр должен быть самым первым, иначе может получиться, что ты загрузил вопрос и ответ, а юзер не залогинен и выходит, зря грузили данные из базы.

File: app/models/answer.rb [hunk 16]
```diff
@@ -3,4 +3,16 @@ class Answer < ActiveRecord::Base
   belongs_to :user
 
   validates :body, :user_id, :question_id, presence: true
+
+  def self.default_scope
```
Comment (vkurennov) [inline app/models/answer.rb]: Лучше дефолтный скоуп так задать:  `default_scope { order(best: :desc, created_at: :desc) }`

File: app/models/answer.rb [hunk 17]
```diff
@@ -3,4 +3,16 @@ class Answer < ActiveRecord::Base
...
+    order(best: :desc, created_at: :desc)
+  end
+
+  def best!
+    ActiveRecord::Base.transaction do
+      question.answers.update_all(best: false)
```
Comment (vkurennov) [inline app/models/answer.rb]: update_all не выбрасывает исключения. Значит, если что-то пойдет не так, то транзакция не откатится. Надо что-то придумать, чтобы при сбросе старого лучшего ответа выбрасывалось исключение.

File: config/routes.rb [hunk 18]
```diff
@@ -1,9 +1,15 @@
 Rails.application.routes.draw do
   devise_for :users
+
   resources :questions do
-    resources :answers, shallow: true
+    resources :answers, shallow: true do
+      member do
```
Comment (vkurennov) [inline config/routes.rb]: Можно вот так, если всего один метод: `post :best, on: :member`

File: spec/models/answer_spec.rb [hunk 19]
```diff
@@ -5,4 +5,18 @@
...
+  let!(:answer2) { create(:answer, question: question) }
+
+  it "make best answer" do
+    answer.best!
+    expect(answer).to be_best
+  end
+
+  it "doesn't make best other answer" do
+    answer.best!
+    expect(answer2).to_not be_best
```
Comment (vkurennov) [inline spec/models/answer_spec.rb:22]: А что, answer2 был лучшим? Что-то не видно, где у него флаг best выставлен в true

**Comment by vkurennov** [comment]
Исправь замечания

File: app/models/answer.rb [hunk 20]
```diff
@@ -4,13 +4,11 @@ class Answer < ActiveRecord::Base
...
 
-  def self.default_scope
-    order(best: :desc, created_at: :desc)
-  end
+  default_scope { order(best: :desc, created_at: :desc) }
 
   def best!
     ActiveRecord::Base.transaction do
-      question.answers.update_all(best: false)
+      question.answers.each { |q| q.update_attributes!(best: false) }
```
Comment (vkurennov) [inline app/models/answer.rb]: Проще было бы обновлять только лучший ответ, он же по идее, один. И используй update! вместо update_attribute

**Comment by vkurennov** [comment]
Принято, но поправь замечания перед выполнением следующего задания
